src/popup/vault/view.component.ts

```
@@ -80,6 +115,89 @@ export class ViewComponent extends BaseViewComponent {
         });
     }
 
+    async fillCipher() {
+
+        if (this.pageDetails == null || this.pageDetails.length === 0) {
+            this.platformUtilsService.showToast('error', null,
+                this.i18nService.t('autofillError'));
+            return false;
+        }
+
+        try {
+            this.totpCode = await this.autofillService.doAutoFill({
+                cipher: this.cipher,
+                pageDetails: this.pageDetails,
+                doc: window.document,
+            });
+            if (this.totpCode != null) {
+                this.platformUtilsService.copyToClipboard(this.totpCode, { window: window });
+            }
+
+            this.platformUtilsService.showToast('success', null,
+                this.i18nService.t('autoFillSuccess'));
+        } catch {
+            this.ngZone.run(() => {
+                this.platformUtilsService.showToast('error', null,
+                    this.i18nService.t('autofillError'));
+                this.changeDetectorRef.detectChanges();
+            });
+            return false;
+        }
+
+        return true;
+    }
+
+    async fillCipherAndSave() {
+        const didAutofill: boolean = await this.fillCipher();
+
+        if (didAutofill) {
+            const tab = await BrowserApi.getTabFromCurrentWindow();
+            if (!tab) {
+                throw new Error('No tab found.');
+            }
+
+            if (this.cipher.login.uris == null) {
+                this.cipher.login.uris = [];
+            } else {
+                if (this.cipher.login.uris.some((uri) => uri.uri === tab.url)) {
+                    this.platformUtilsService.showToast('error', null,
+                        this.i18nService.t('uriExists'));
+                    return;
+                }
```

Not sure we even need to show a message here. You could just return and silently do nothing.